exporter/opentelemetry-exporter-prometheus-remote-write/src/opentelemetry/exporter/prometheus_remote_write/__init__.py

```diff
@@ -142,50 +153,144 @@ def export(
+
+    def _convert_from_histogram(
         self, histogram_record: ExportRecord
     ) -> TimeSeries:
-        raise NotImplementedError()
-
-    def convert_from_last_value(
+        count = 0
+        timeseries = []
+        for bound in histogram_record.aggregator.checkpoint.keys():
+            bound_str = "+Inf" if bound == float("inf") else str(bound)
+            value = histogram_record.aggregator.checkpoint[bound]
+            timeseries.append(
+                self._create_timeseries(
+                    histogram_record,
+                    histogram_record.instrument.name,
+                    value,
+                    labels=[("le", bound_str)],
+                )
+            )
+            count += value
+        name = histogram_record.instrument.name + "_count"
+        timeseries.append(
+            self._create_timeseries(histogram_record, name, float(count))
+        )
+        return timeseries
+
+    def _convert_from_last_value(
         self, last_value_record: ExportRecord
     ) -> TimeSeries:
-        raise NotImplementedError()
-
-    def convert_from_value_observer(
+        return [
+            self._create_timeseries(
+                last_value_record,
+                last_value_record.instrument.name,
+                last_value_record.aggregator.checkpoint,
+            )
+        ]
+
+    def _convert_from_value_observer(
         self, value_observer_record: ExportRecord
     ) -> TimeSeries:
-        raise NotImplementedError()
+        timeseries = []
+        for agg_type in ["min", "max", "sum", "count", "last"]:
+            timeseries.append(
+                self._create_timeseries(
+                    value_observer_record,
+                    value_observer_record.instrument.name + "_" + agg_type,
+                    getattr(
+                        value_observer_record.aggregator.checkpoint, agg_type
+                    ),
+                )
+            )
+        return timeseries
 
-    def convert_from_quantile(
+    # TODO: Implement convert from quantile once supported by SDK for Prometheus Summaries
+    def _convert_from_quantile(
         self, summary_record: ExportRecord
     ) -> TimeSeries:
         raise NotImplementedError()
 
     # pylint: disable=no-member
-    def create_timeseries(
-        self, export_record: ExportRecord, name, value: float
+    def _create_timeseries(
+        self, export_record: ExportRecord, name, value: float, labels=None
     ) -> TimeSeries:
-        raise NotImplementedError()
-
-    def create_sample(self, timestamp: int, value: float) -> Sample:
-        raise NotImplementedError()
-
-    def create_label(self, name: str, value: str) -> Label:
-        raise NotImplementedError()
+        timeseries = TimeSeries()
+        seen = set()
+
+        def add_label(label_name, label_value):
+            # Label name must contain only alphanumeric characters and underscores
+            label_name = re.sub("[^\\w_]", "_", label_name)
```

Comment:
I'm wondering if we might get labels that are supposed to be unique but will be dropped due to removal of these characters. Is the alphanumeric + underscores limitation of the labels specific for Prometheus?

e.g. 
`testing` and `${testing}`

Reply:
The limitation is specific to the TimeSeries type https://prometheus.io/docs/concepts/data_model/ . You bring up a valid point. Since the non-alphanumeric characters are replaced with underscores: '$testing' and '_testing' will be considered the same label which would be an issue. Also, a label_name cannot start with two underscores. Would it be better to validate label names (raise Error) or add more complex sanitizing? I am not quite sure how to prevent duplicates when addressing label names that contain non-alphanumeric characters.

Reply:
If the limitation is Prometheus specific, I am okay with the current implementation. The only problem was if it were a limitation we imposed on the OT side, then Prometheus users would have unexpected behaviour. But if it's on the Prometheus side, then user's should EXPECT this behaviour.

I wonder in this case if we should DROP labels that defy the `TimeSeries` definition (and maybe log a warning), instead of only dropping the duplicate.

Reply:
In the rare case this occurs, I think whoever is misusing the exporter will value the label with the non-ascii characters as much as the valid one, so I suggest dropping the duplicate but logging a warning that after sanitizing, a duplicate label had to be dropped.